[tadacheng] iP

tadacheng opened the pull request:
### 🦆 Duck

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Duck is a chat bot that helps you stay on top of your tasks effortlessly. It's,

- _Text-based_
- ~Complicated~ **Simple** to use
- Lightweight and fast

Getting Started 

1. Download [here](https://github.com/tadacheng/ip)
2. Install and launch it
3. Add your tasks
4. Stay organised

Features:

- [x]  Managing tasks 
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

tadacheng committed:
Level 0. Rename, Greet, Exit

tadacheng committed:
Level 1. Echo

tadacheng committed:
Level 2. Add, List

tadacheng committed:
Add Task

tadacheng committed:
Level 3. Mark as Done

tadacheng committed:
Add todo, deadline, event

tadacheng committed:
Level 4. ToDos, Events, Deadlines

tadacheng committed:
Added Text UI Testing

tadacheng committed:
Added DuckException

tadacheng committed:
Added Error Handling

tadacheng committed:
Level 6. Delete

tadacheng committed:
Remove redundant Strings

tadacheng committed:
Update change variable name to follow coding standards

tadacheng committed:
Add loadTasksFromSave

tadacheng committed:
Add toFileFormat to Tasks

tadacheng committed:
Add saveTasksToFile

tadacheng committed:
Change import to not use wildcard

tadacheng committed:
Add duck.txt to gitignore

tadacheng committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Change import to not use wildcard
  Add saveTasksToFile
  Add toFileFormat to Tasks
  Add loadTasksFromSave
  Update change variable name to follow coding standards

tadacheng committed:
Add save functionality to task list

Let's
* implement automatic saving of tasks to a file on the hard disk whenever the task list changes.
* add function to load tasks from the file when the program starts.
* create directories and files if they do not exist.

tadacheng committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Add save functionality to task list

tadacheng committed:
Change dates in Deadline and Event from String to LocalDateTime

tadacheng committed:
Add LocalDateTime support for Deadline and Event tasks

* Updated Deadline and Event classes to store and handle dates and times using java.time.LocalDateTime
* Implemented parsing for input date and time in "yyyy-MM-dd HHmm" format
* Enhanced toString methods to display dates in a user-friendly format

tadacheng committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Add LocalDateTime support for Deadline and Event tasks
  Change dates in Deadline and Event from String to LocalDateTime

tadacheng committed:
Move Tasklist operations to separate class

tadacheng committed:
Move storage functions to separate class

tadacheng committed:
Move UI elements to a separate class

tadacheng committed:
Move parsing of user commands to separate class
Add Command classes for each command

tadacheng committed:
Refactor code to implement OOP principles

Extracted
* Ui elements that deals with interactions with the user
* Storage functions that deals with loading tasks from the file and saving tasks in the file
* Parser to deal with making sense of the user command
* TaskList contains the task list and has operations to add/delete tasks in the list

tadacheng committed:
Merge branch 'branch-MoreOOP'

* branch-MoreOOP:
  Refactor code to implement OOP principles
  Move parsing of user commands to separate class Add Command classes for each command
  Move UI elements to a separate class
  Move storage functions to separate class
  Move Tasklist operations to separate class

tadacheng committed:
Divide classes into packages

Move
* class that inherit Command to command
* class that inherit Task to task
* class with user interaction to ui

tadacheng committed:
Merge branch 'branch-Packages'

* branch-Packages:
  Divide classes into packages

tadacheng committed:
Merge branch 'add-gradle-support'

* add-gradle-support:
  build.gradle: Prevent generating a second JAR file
  Bump gradle and lib version
  Add Gradle support

tadacheng committed:
Add Gradle support

tadacheng committed:
Add unit testing using JUnit

Test
* Parser
* Storage
* TaskList

tadacheng committed:
Merge branch 'branch-JUnit'

* branch-JUnit:
  Add unit testing using JUnit

tadacheng committed:
Add JavaDocs to TaskList and Parser

tadacheng committed:
Add javadocs to Command classes

tadacheng committed:
Add javadocs for Task Classes

tadacheng committed:
Add Javadocs for Duck application

tadacheng committed:
Add modification to fit Coding Style

Fix
* Spacings
* Imports

tadacheng committed:
Add Find functionality to Tasklist

Add
* New FindCommand
* Update to Parser
* Update to TaskList with findTask function

tadacheng committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add Javadocs for Duck application
  Add javadocs for Task Classes
  Add javadocs to Command classes
  Add JavaDocs to TaskList and Parser

tadacheng committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Add modification to fit Coding Style

# Conflicts:
#	src/main/java/duck/exception/DuckException.java
#	src/main/java/duck/task/Todo.java
#	src/main/java/duck/ui/TaskList.java

tadacheng committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add Find functionality to Tasklist

# Conflicts:
#	src/main/java/duck/task/Task.java

tadacheng committed:
Merge branch 'master' into branch-A-JavaDoc

* master:
  Add Find functionality to Tasklist
  Add modification to fit Coding Style

tadacheng committed:
Add Javadocs to Find functionality

tadacheng committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add Javadocs to Find functionality

tadacheng committed:
Set up check style on Gradle and Intellij

Scan and fix some check style problems

tadacheng committed:
Merge branch 'branch-CheckStyle'

* branch-CheckStyle:
  Set up check style on Gradle and Intellij

tadacheng committed:
Set up JavaFX to work with Gradle

Add
* Launcher
* Main

tadacheng committed:
Follow JavaFX tutorial for Gui

Added features up to part 4

tadacheng committed:
Add Gui support for command execution

* Add UI elements for command input and execution
* Connect command handling logic to GUI components

tadacheng committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Add Gui support for command execution
  Follow JavaFX tutorial for Gui
  Set up JavaFX to work with Gradle

Hugo-Chia commented on src/main/java/duck/exception/DuckException.java:
```diff
@@ -0,0 +1,17 @@
+package duck.exception;
```
Maybe it is possible handle other kinds of exceptions here?

Hugo-Chia commented on src/main/java/duck/command/MarkCommand.java:
```diff
@@ -0,0 +1,47 @@
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws DuckException {
+        if (this.id < 0 || this.id >= tasks.size()) {
+            throw new DuckException("Invalid task number. Use list to view task id");
```
Could we use "Use list command ..." or something similar instead, so the user would better understand the error

Hugo-Chia commented on src/main/java/duck/command/Command.java:
```diff
@@ -0,0 +1,49 @@
+ */
+public abstract class Command {
+    protected static final String CMD_NOT_EXE_RESPONSE = "Command has not been executed";
+    protected boolean hasExecuted = false;
```
Great job on the boolean variable naming! All of them are phrased in "question-style", hasExecuted, isExit, etc.

Hugo-Chia reviewed (commented):
Overall code is impressive! I could not find any coding violations and code was very simple to understand. Classes were well separated, methods and variables were well-named too! Great job! :)

Seethevon commented on src/main/java/duck/ui/MainWindow.java:
```diff
@@ -0,0 +1,57 @@
+ * Controller for the main GUI.
+ * Taken from https://se-education.org/guides/tutorials/javaFxPart4.html
+ */
+public class MainWindow extends AnchorPane {
```
I like how your class names are nouns and your methods are verbs. Good job on that 👍 .

Seethevon commented on src/main/java/duck/ui/MainWindow.java:
```diff
@@ -0,0 +1,57 @@
+     * the dialog container. Clears the user input after processing.
+     */
+    @FXML
+    private void handleUserInput() {
```
I like that the method name explains what the method does and is not vague. It makes the overall code easily readable 😄.

tadacheng committed:
Add assertions to prevent unexpected parameters

The code previously lacked assertions, which could lead to unexpected
errors and invalid states going unnoticed during development.

A more defensive approach with assertions helps catch bugs early
and ensures that critical assumptions hold.

Added assert to
* Parser and TaskList to prevent null inputs
* Storage to ensure files are created

tadacheng committed:
Improve code quality by deleting unused classes

The code previously used a terminal for the ui.
With the new JavaFX it is now not needed anymore.

tadacheng committed:
Refactor code to use java streams in TaskList and Storage

Use .stream().filter() in findTasks to simplify task searching.
Use .stream().map().forEach() in save to simplify task writing.

tadacheng committed:
Merge pull request #1 from tadacheng/branch-A-Assertions

Add assertions to prevent unexpected parameters

tadacheng committed:
Merge branch 'master' into branch-A-CodeQuality

tadacheng committed:
Merge pull request #2 from tadacheng/branch-A-CodeQuality

Improve code quality by deleting unused classes

tadacheng committed:
Merge branch 'master' into branch-A-Streams

tadacheng committed:
Merge pull request #3 from tadacheng/branch-A-Streams

Refactor code to use java streams in TaskList and Storage

tadacheng committed:
Add Recurring Task

The task will be auto create a new task when it is mark done

tadacheng committed:
Merge branch 'branch-BCD-Extension'

* branch-BCD-Extension:
  Add Recurring Task

tadacheng committed:
Set up CI

tadacheng committed:
Merge branch 'branch-A-CI'

* branch-A-CI:
  Set up CI

tadacheng committed:
Fix Tests to handle added assertions

tadacheng committed:
Fix error with test of no parent file

tadacheng committed:
Fix os dependent file path

tadacheng committed:
Modify test to handle windows

tadacheng committed:
Add error handling when date time are not in correct format

tadacheng committed:
Merge branch 'branch-A-MoreErrorHandling'

* branch-A-MoreErrorHandling:
  Add error handling when date time are not in correct format

tadacheng committed:
Add testing for Duck

tadacheng committed:
Merge branch 'branch-A-MoreTesting'

* branch-A-MoreTesting:
  Add testing for Duck

tadacheng committed:
Add User Guide

tadacheng committed:
Merge branch 'branch-A-UserGuide'

* branch-A-UserGuide:
  Add User Guide

tadacheng committed:
Fix header comments

tadacheng committed:
Rename UI.png to Ui.png

tadacheng committed:
Update README.md
